### ISSUE TITLE
Add instructions for building mfoc on Windows.

### ISSUE THREAD

**Issue by DavidBerdik** [nfc-tools/mfoc#74]
The available documentation for mfoc details how to build it on *nix-based systems, but not Windows. Even if this project is not intended for Windows, I believe the project could still benefit from providing instructions for building for Windows for those who would like to try it.

Unfortunately, using a VM with USB passthrough is not possible, as the program does not work properly in a VM, and USB passthrough is apparently not supported by the Windows Subsystem for Linux for anything other than external drives and flash drives.

**Comment by xavave** [comment]
Hi,
I've cross-compiled mfoc and nfc utils for windows 64 bits , you can find them on my blog:
~~http://legacy.averbouch.biz/libnfc-and-nfc-utils-binaries-on-windows-10/~~
https://blog.avxt.fr/libnfc-and-nfc-utils-binaries-on-windows-10/

**Comment by DavidBerdik** [comment]
Hello @xavave,

Thank you so much for sharing this! This is wonderful! Do you by chance have any documentation on what you changed from the original Linux version? Or is it linked in your post and I just didn't see it? I ask because I would be interested in trying to maintain a Windows version that tries to keep up to date with changes applied to the Linux version. You certainly provide a wonderful starting point for doing just that.

**Comment by xavave** [comment]
@DavidBerdik I used libnfc1.7.1 and mainly used tutorial here : https://github.com/peacepenguin/libnfc-unofficialbuilds

btw I've also cross-compiled libfreefare for windows 64, you can find binaries in attachment at the end of my post here : 
https://github.com/nfc-tools/libfreefare/pull/52 (post named : allow to cross-compile 32 and 64 bit dlls using MinGW)

**Comment by DavidBerdik** [comment]
After further experimentation, I am not sure that the `mfcuk` prebuilt binary that you provide works properly. I've had it running an attack on a card for several hours now, and it doesn't look like it's making any progress. I've broken this same card using the same reader using mfcuk on Ubuntu and it's worked, so I do not really no what to make of this.

```
-----------------------------------------------------
Let me entertain you!
    uid: b2588311
   type: 08
    key: 000000000000
  block: 03
diff Nt: 19905
   hit4: 0
  auths: 19905
-----------------------------------------------------
```

My command to start running the attack was `mfcuk_keyrecovery_darkside.exe -C -R -1:A -s 250 -S 250 -v 3 -o mfcuk-dump`.

Any suggestions @xavave?

**Comment by xavave** [comment]
@DavidBerdik Did you use exatly the same command parameters when it worked on Ubuntu ?

**Comment by DavidBerdik** [comment]
@xavave Yes, I used exactly the same command parameters when testing on Ubuntu. I can say that with absolute confidence because I copied the command off to a text file for future use when testing on Ubuntu.

As for which version I used, I am not sure because I did not compile from source. I installed it using `apt`.

I will try testing now and let you know how it goes.

**Comment by xavave** [comment]
@DavidBerdik I’m not sure the warnings make a difference. 

I saw an old “official” windows version here.
https://github.com/nfc-tools/mfcuk/releases/tag/windows
Do you see a difference on testing compared to my releases ?

**Comment by DavidBerdik** [comment]
Unfortunately, none of the `mfcuk` variations you have made available seem to work for me. As for sharing the dump of my card, I can (it's only an old hotel room key, so there aren't any security concerns), but you may have to wait a while. The hard drive in the laptop that I used to do this on died a month or two ago. That hard drive had the only copy of the card dump that I made, and I haven't had a chance to do anything about the laptop, and as you likely know, using USB pass-through to an Ubuntu VM does not cooperate with the ACR122U.

**Comment by DavidBerdik** [comment]
@xavave  Unfortunately, none of the `mfcuk` variations you have made available seem to work for me. As for sharing the dump of my card, I can (it's only an old hotel room key, so there aren't any security concerns), but you may have to wait a while. The hard drive in the laptop that I used to do this on died a month or two ago. That hard drive had the only copy of the card dump that I made, and I haven't had a chance to do anything about the laptop, and as you likely know, using USB pass-through to an Ubuntu VM does not cooperate with the ACR122U.

**Comment by xavave** [comment]
@DavidBerdik on my laptop When I test my l’exil I can see different « diff nt » and « auth » values. So I don’t understand the problem for now. I can wait for your dump, no problem. Thank you

**Comment by xavave** [comment]
@DavidBerdik I tried also mfcuk on linux but I get the same results : same count of auths and diff nt , maybe the issue comes with the reader ACR122U-A9 ? I have no clue for now

**Comment by DavidBerdik** [comment]
@xavave I expect that I will have time to play with this a little tomorrow. I will try to take a dump of it on Linux and post it here. I remember the last time I cracked the card, I used `mfcuk` to crack all of the A keys which took a great deal of time. I hadn't realized at the time that you only need one of the A keys to derive the rest. Hopefully that will speed things up a bit.

**Comment by DavidBerdik** [comment]
@xavave An update for you: I am trying to crack the card on my Linux laptop right now, and it's doing the same thing. I know it worked before so I'm going to try to just let it run its course. I'm at about 37,000 auths right now.

**Comment by xavave** [comment]
@DavidBerdik do you use exactly the same card reader than before ?

**Comment by DavidBerdik** [comment]
@xavave Yep! Same card reader and same card. I woke up this morning to find that `mfcuk` had crashed, so I am trying again with a different card but the same reader.

**Comment by DavidBerdik** [comment]
@xavave Have you had any luck? I have been trying with another, different card now for several days and it's doing the same thing. I have also cracked this card before successfully.

**Comment by xavave** [comment]
@DavidBerdik I didn't try again, but you are saying that on linux and, on windows mfcuk doesn't work ? and on linux it was working previously ? I ha ve no clue, or mfcuk has been upgraded meanwhile and has new bugs ?

**Comment by DavidBerdik** [comment]
@xavave I finally managed to get a key by setting `-R 2:A` when attacking on Linux, but I still have not had any success on Windows. So far, I have only tested with this parameter using the original `mfcuk` that you linked to in your initial reply. I am going to try the other builds as well though.

**Comment by xavave** [comment]
@DavidBerdik good 👍🏻
How long did it take on Linux ? Could you send me the dump please ?

**Comment by DavidBerdik** [comment]
@xavave It took about 30 minutes. Unfortunately, the card that I successfully dumped was not the hotel room key that I told you about earlier. The card that I managed to get working is actually sensitive, so I cannot share it. I am going to keep experimenting with the other card on Linux to see if I can get a dump to share with you.

**Comment by xavave** [comment]
@DavidBerdik  ok 👍 thank you

**Comment by DavidBerdik** [comment]
@xavave It just occurred to me that even though `mfcuk` testing with the hotel room key is not working, I can still give you an `mfoc` dump since the card uses default keys, so I have attached one. I am still going to keep trying to get `mfcuk` working on it though. Thanks for your help with this!

[mfoc_hotel_room_key_dump.zip](https://github.com/nfc-tools/mfoc/files/4145301/mfoc_hotel_room_key_dump.zip)

**Comment by xavave** [comment]
thank you, I'm going to test it now

**Comment by DavidBerdik** [comment]
@xavave Excellent! Good luck!

**Comment by xavave** [comment]
@DavidBerdik after a few minutes, diff nt and auths or different : 
![image](https://user-images.githubusercontent.com/3501675/73616530-e005fb00-4614-11ea-916f-ada0b37d366e.png)
![image](https://user-images.githubusercontent.com/3501675/73616532-e98f6300-4614-11ea-9354-183a08337cfe.png)
btw I used key B in parameters for testing: -C -R -1:B -s 250 -S 250 -v 3 -o mfcuk-dump


the mfcuk I used is attached here
[mfcuk_keyrecovery_darkside.zip](https://github.com/nfc-tools/mfoc/files/4145582/mfcuk_keyrecovery_darkside.zip)

I will let you know how it ends..

**Comment by DavidBerdik** [comment]
@xavave I get different values for `diff Nt` and `auths` after a few minutes too, but it still doesn't seem to work for me. I am trying again on Ubuntu using `-R 2:A`, and here is where I am as of a few moments ago.

![Screenshot from 2020-02-02 18-14-42](https://user-images.githubusercontent.com/7663225/73617011-7a037e80-45e8-11ea-8dd7-4cebe86ce05c.png)

I was actually not aware that you could attack for B keys using `mfcuk`. I thought it was only for getting at least one A key which you then used with `mfoc` to get the B keys as well as the remaining A keys. I will let it continue running on Ubuntu for now, and if it does not crack a key within 24 hours, I will give up and try something else.

**Comment by xavave** [comment]
@DavidBerdik I was reading on some [chinese forums (that I've google translated :)](https://translate.google.com/translate?sl=auto&tl=en&u=www.mobile01.com%2Ftopicdetail.php%3Ff%3D634%26t%3D5938144)  ) that the results depends on the sector you attack, so many tries can be done before finding

**Comment by DavidBerdik** [comment]
@xavave Thanks for sharing! I have seen similar claims on other websites, but as I understand, my attacks are taking longer than they should have to. In several cases over the last few weeks, I let them run for over 24 hours without any progress made at all. In some of those cases, `mfcuk` would start throwing an error with each read attempt. Unfortunately I did not document what the error was so I cannot share, but I think it was a bug with the program itself and not an issue with the card.

**Comment by xavave** [comment]
@DavidBerdik it doesn't seem to work for me neither 
![image](https://user-images.githubusercontent.com/3501675/73646001-c5fe0400-4678-11ea-9b02-8663b61263f0.png) I saw there is a dedicated support for mfcuk btw
https://github.com/nfc-tools/mfcuk/issues/39
I wonder if it comes from ACR122 tag reader on windows version, but, I dont't have other USB reader to test for now

**Comment by DavidBerdik** [comment]
@xavave Have you experimented with doing it on Linux at all? Also, why do some versions of `mfcuk` have a `hit4` field while others do not? What is that all about? I assume they are different versions of the program, but I do not know what it means. (Well, I actually do not know what any of those fields are except for the `uid` and `auths` ones.)

**Comment by xavave** [comment]
@DavidBerdik I should try on Linux again. Tonight I let mfcuk run with libnfc.dll v1.5 to check if issue can be related to latest Li nfc version 1.7. 
Hit4 is the number of successful auths as you can see in dr skottky mfcuk improved version: 
![4A05CED6-7842-47F7-98ED-BE8291AA9B4C](https://user-images.githubusercontent.com/3501675/73698836-87049880-46e2-11ea-98cb-10241d0ec0a3.png)

**Comment by xavave** [comment]
maybe some cards are not sentitive to mfcuk: http://www.cs.ru.nl/~rverdult/Ciphertext-only_Cryptanalysis_on_Hardened_Mifare_Classic_Cards-CCS_2015.pdf

![image](https://user-images.githubusercontent.com/3501675/73737242-8b609e00-4742-11ea-98b4-d5c81d202d88.png)

**Comment by xavave** [comment]
@DavidBerdik I've decommented debug print in mfcuk source code on windows and ubuntu and, i've noticed that on ubuntu, nonce distance is between 0 and about 65535 and values follows each others
![image](https://user-images.githubusercontent.com/3501675/73777989-b969d080-478a-11ea-81e3-6c060519869f.png) diff nt much lower than auths , unlike on windows
on windows this is not working properly : I never see nonce distance lower than a few thousands
anyway , I run mfcuk on ubuntu for more than 1 hour with your dump written on a card , and also no successful result

**Comment by DavidBerdik** [comment]
I have had an `mfcuk` attack running against the card for 2 days now and still nothing has come of it. I am not really sure what to do at this point. I have another card that I can try that I had successfully cracked with `mfcuk` before. Unfortunately though, it is sensitive and therefore I cannot share the dump.

Also, I am not sure what the nonce distance value is supposed to mean.

**Comment by DavidBerdik** [comment]
@xavave I'm trying the other card that I mentioned in my previous post, and I am seeing no progress. The `diff Nt` and `auths` values are just incrementing together.

![Screenshot from 2020-02-04 21-54-20](https://user-images.githubusercontent.com/7663225/73808227-30a86000-479d-11ea-95b2-bd84ff92a202.png)

**Comment by xavave** [comment]
@DavidBerdik some cards are protected against dark side attack. But I still can’t explain why different  mfcuk versions never found a key for me neither on Ubuntu neither on windows with different cards including blank ones. I’ve  just ordered a proxmark 3 Tag reader so I will be able to compare results with Acr122u-a9 soon.

**Comment by DavidBerdik** [comment]
@xavave What disturbs me is that I have already cracked this card using `mfcuk` on Ubuntu but now it's not working. Furthermore, this is the same card that I cracked 2:A on last week, but now it's acting weird again. I don't understand it.

As for how long it took when I did it last year, it was a several-hour process.

**Comment by xavave** [comment]
@DavidBerdik I've reopened this issue on dedicated mfcuk discussion : https://github.com/nfc-tools/mfcuk/issues/76

**Comment by GitH-Michael** [comment]
> Hi, I've cross-compiled mfoc and nfc utils for windows 64 bits , you can find them on my blog: http://legacy.averbouch.biz/libnfc-and-nfc-utils-binaries-on-windows-10/

Yo bro, apparently the website you linked is empty! could you please share this somewhere else? I would really appreciate it

**Comment by DavidBerdik** [comment]
> Yo bro, apparently the website you linked is empty! could you please share this somewhere else? I would really appreciate it

The Internet Archive has it: https://web.archive.org/web/20230331102618/http://legacy.averbouch.biz/libnfc-and-nfc-utils-binaries-on-windows-10/

**Comment by xavave** [comment]
Hi @GitH-Michael @DavidBerdik,
Sorry for the late reply!
I've just got my website back up and running on a new domain
The attachment upload feature is not back yet, but I'll try to get it back up and running in the next few days
Here is the link: https://blog.avxt.fr/libnfc-and-nfc-utils-binaries-on-windows-10/

**Comment by DavidBerdik** [comment]
@xavave Unfortunately it does not seem to be loading for me.

**Comment by xavave** [comment]
@DavidBerdik did you try with https://blog.avxt.fr ?
do you see an error message when trying to load ?
![image](https://github.com/user-attachments/assets/37d2fced-5aca-491b-8866-5f8d07240250)

**Comment by DavidBerdik** [comment]
@xavave This is what I see.

![image](https://github.com/user-attachments/assets/29c52d4b-a1cb-4cb3-bb49-b27d522c5782)
